### PR TITLE
segcompaction featuring verticalcompaction

### PR DESCRIPTION
**Pull request by freemandealer**
This patchset applies the following changes:
- using vertical compaction machanism to do segcompaction
- basic (WIP) refraction to separate segcompaction logic from BetaRowsetWriter
- add segcompaction specific ut and regression tests

# Proposed changes

Issue Number: close #xxx

## Problem summary

Describe your changes.

## Checklist(Required)

* [ ] Does it affect the original behavior
* [ ] Has unit tests been added
* [ ] Has document been added or modified
* [ ] Does it need to update dependencies
* [ ] Is this PR support rollback (If NO, please explain WHY)

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: be/src/olap/rowset/beta_rowset_writer.cpp [hunk 1]
```diff
@@ -116,9 +118,11 @@ Status BetaRowsetWriter::add_block(const vectorized::Block* block) {
     return _add_block(block, &_segment_writer);
 }
 
-RowwiseIteratorUPtr BetaRowsetWriter::_get_segcompaction_reader(
-        SegCompactionCandidatesSharedPtr segments, std::shared_ptr<Schema> schema,
-        OlapReaderStatistics* stat, uint64_t* merged_row_stat) {
+std::unique_ptr<vectorized::VerticalBlockReader> BetaRowsetWriter::_get_segcompaction_reader(
```
Comment (yixiutt) [inline be/src/olap/rowset/beta_rowset_writer.cpp:121]: not return unique ptr, use std::unique_ptr<vectorized::VerticalBlockReader>* in input params

File: be/src/olap/rowset/beta_rowset_writer.cpp [hunk 2]
```diff
@@ -133,18 +137,21 @@ RowwiseIteratorUPtr BetaRowsetWriter::_get_segcompaction_reader(
...
+
+    TabletReader::ReaderParams reader_params;
+    reader_params.is_segcompaction = true;
+    reader_params.segment_iters_ptr = &seg_iterators;
+    // no reader_params.version shouldn't break segcompaction
+    reader_params.tablet_schema = _context.tablet_schema;
+    reader_params.tablet = tablet;
+    reader_params.return_columns = return_columns;
+    reader_params.is_key_column_group = is_key;
+    reader->init(reader_params);
```
Comment (yixiutt) [inline be/src/olap/rowset/beta_rowset_writer.cpp:152]: need check status?

File: be/src/olap/merger.cpp [hunk 3]
```diff
@@ -112,7 +110,7 @@ Status Merger::vmerge_rowsets(TabletSharedPtr tablet, ReaderType reader_type,
     }
 
     RETURN_NOT_OK_LOG(
-            dst_rowset_writer->flush(),
+            dst_rowset_writer->flush_by_merger(),
```
Comment (yixiutt) [inline be/src/olap/merger.cpp:113]: flush not ok?

File: be/src/vec/olap/vertical_block_reader.cpp [hunk 4]
```diff
@@ -86,9 +86,25 @@ Status VerticalBlockReader::_init_collect_iter(const ReaderParams& read_params)
...
-    RETURN_IF_ERROR(
-            _get_segment_iterators(read_params, &segment_iters, &iterator_init_flag, &rowset_ids));
-    CHECK(segment_iters.size() == iterator_init_flag.size());
+    std::vector<RowwiseIteratorUPtr>* segment_iters_ptr = read_params.segment_iters_ptr;
+
+    if (!segment_iters_ptr) {
+        RETURN_IF_ERROR(_get_segment_iterators(read_params, &segment_iters, &iterator_init_flag,
+                                               &rowset_ids));
+        CHECK(segment_iters.size() == iterator_init_flag.size());
+        segment_iters_ptr = &segment_iters;
```
Comment (yixiutt) [inline be/src/vec/olap/vertical_block_reader.cpp:95]: pass segment_iters_ptr to _get_segment_iterators?

**Comment by freemandealer** [comment]
run buildall

**Comment by hello-stephen** [comment]
TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 33.91 seconds
 stream load tsv:          465 seconds loaded 74807831229 Bytes, about 153 MB/s
 stream load json:         36 seconds loaded 2358488459 Bytes, about 62 MB/s
 stream load orc:          67 seconds loaded 1101869774 Bytes, about 15 MB/s
 stream load parquet:          28 seconds loaded 861443392 Bytes, about 29 MB/s
 https://doris-community-test-1308700295.cos.ap-hongkong.myqcloud.com/tmp/20230224092145_clickbench_pr_103969.html

**Review by zhannngchen — DISMISSED** [review]
LGTM

**Comment by freemandealer** [comment]
run buildall

**Comment by freemandealer** [comment]
run buildall

**Comment by freemandealer** [comment]
run p0

**Review by zhannngchen — APPROVED** [review]
LGTM

**Pull request merged**